presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.sanity.warnings;
+
+import io.prestosql.Session;
+import io.prestosql.execution.warnings.WarningCollector;
+import io.prestosql.metadata.FunctionMetadata;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.ResolvedFunction;
+import io.prestosql.spi.PrestoWarning;
+import io.prestosql.sql.planner.ExpressionExtractor;
+import io.prestosql.sql.planner.SimplePlanVisitor;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+import io.prestosql.sql.planner.sanity.PlanSanityChecker;
+import io.prestosql.sql.tree.DefaultTraversalVisitor;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.FunctionCall;
+
+import java.util.Optional;
+
+import static io.prestosql.spi.connector.StandardWarningCode.DREPRCATED_FUNCTION;
+
+public final class DeprecatedFunctionWarning
+        implements PlanSanityChecker.Checker
+{
+    @Override
+    public void validate(PlanNode plan, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider types, WarningCollector warningCollector)
+    {
+        for (Expression expression : ExpressionExtractor.extractExpressions(plan)) {
+            new DefaultTraversalVisitor<Void, Void>()
+            {
+                @Override
+                protected Void visitFunctionCall(FunctionCall node, Void context)
+                {
+                    Optional<ResolvedFunction> resolvedFunction = ResolvedFunction.fromQualifiedName(node.getName());
+                    if (resolvedFunction.isPresent()) {
+                        FunctionMetadata functionMetadata = metadata.getFunctionMetadata(resolvedFunction.get());
+                        if (functionMetadata.isDeprecated()) {
+                            warningCollector.add(new PrestoWarning(DREPRCATED_FUNCTION,
+                                    String.format("Detected use of deprecated function: %s", functionMetadata.getSignature().getName())));
```

Comment:
Ideally, the deprecation message should be able to contain instructions about the reason for deprecation and possible substitutions. So, we'll need to capture this information in the deprecation annotation. Since `@Deprecated` doesn't take a description, we might need an additional annotation.

Reply:
This requires adding a new annotation and pass it into FunctionMetadata, which looks like a bit overkill to me. The existing `@Description` can be used to explain the substitution, just like what we do in session description:
https://github.com/prestosql/presto/blob/5d013879f430fcb90cf37752c501df16baa073fa/presto-main/src/main/java/io/prestosql/SystemSessionProperties.java#L301

Reply:
That's a good point. Let's use that.